broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java

```diff
@@ -0,0 +1,498 @@
+                if (newOffset != halfOffset) {
+                    transactionalMessageBridge.updateConsumeOffset(messageQueue, newOffset);
+                }
+                long newOpOffset = calculateOpOffset(doneOpOffset, opOffset);
+                if (newOpOffset != opOffset) {
+                    transactionalMessageBridge.updateConsumeOffset(opQueue, newOpOffset);
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+            log.error("Check error", e);
+        }
+
+    }
+
+    private long getImmunityTime(String checkImmunityTimeStr, long transactionTimeout) {
+        long checkImmunityTime;
+
+        checkImmunityTime = getLong(checkImmunityTimeStr);
+        if (-1 == checkImmunityTime) {
+            checkImmunityTime = transactionTimeout;
+        } else {
+            checkImmunityTime *= 1000;
+        }
+        return checkImmunityTime;
+    }
+
+    /**
+     * Read op message, parse op message, and fill removeMap
+     *
+     * @param removeMap Half message to be remove, key:halfOffset, value: opOffset.
+     * @param opQueue Op message queue.
+     * @param pullOffsetOfOp The begin offset of op message queue.
+     * @param miniOffset The current minimum offset of half message queue.
+     * @param doneOpOffset Stored op messages that have been processed.
+     * @return Op message result.
+     */
+    private PullResult fillOpRemoveMap(HashMap<Long, Long> removeMap,
+        MessageQueue opQueue, long pullOffsetOfOp, long miniOffset, List<Long> doneOpOffset) {
+        PullResult pullResult = pullOpMsg(opQueue, pullOffsetOfOp, 32);
+        if (null == pullResult) {
+            return null;
+        }
+        if (pullResult.getPullStatus() == PullStatus.OFFSET_ILLEGAL
+            || pullResult.getPullStatus() == PullStatus.NO_MATCHED_MSG) {
+            log.warn("The miss op offset={} in queue={} is illegal, pullResult={}", pullOffsetOfOp, opQueue,
+                pullResult);
+            transactionalMessageBridge.updateConsumeOffset(opQueue, pullResult.getNextBeginOffset());
+            return pullResult;
+        } else if (pullResult.getPullStatus() == PullStatus.NO_NEW_MSG) {
+            log.warn("The miss op offset={} in queue={} is NO_NEW_MSG, pullResult={}", pullOffsetOfOp, opQueue,
+                pullResult);
+            return pullResult;
+        }
+        List<MessageExt> opMsg = pullResult.getMsgFoundList();
+        if (opMsg == null) {
+            log.warn("The miss op offset={} in queue={} is empty, pullResult={}", pullOffsetOfOp, opQueue, pullResult);
+            return pullResult;
+        }
+        for (MessageExt opMessageExt : opMsg) {
+            Long queueOffset = getLong(new String(opMessageExt.getBody(), TransactionalMessageUtil.charset));
+            log.info("Topic: {} tags: {}, OpOffset: {}, HalfOffset: {}", opMessageExt.getTopic(),
+                opMessageExt.getTags(), opMessageExt.getQueueOffset(), queueOffset);
```

Comment:
queueOffset  = getLong(new String(opMessageExt.getBody(), TransactionalMessageUtil.charset)); 
why this  can get the value of HalfOffset？

Reply:
Details can be found in TransactionalMessageService#deletePrepareMessage method.